### ISSUE TITLE
reftable in jgit doesn't respect changes outside of jgit

### ISSUE THREAD

**Issue by zotttelbart** [eclipse-jgit/jgit#102]
### Version

7.0.0

### Operating System

Windows

### Bug description

Changes on reftable outside of jgit (example gitbash) are not visible for jgit

### Actual behavior

`		Git git = Git.init().setDirectory(new File("c:\\temp\\jgit")).call();
		
		git.remoteAdd().setUri(new URIish("https://github.com/eclipse-jgit/jgit.git")).setName("origin").call();
		git.fetch().setRemote("origin").setProgressMonitor(new TextProgressMonitor()).call();
		git.checkout().setName("master").setStartPoint("origin/master").setCreateBranch(true).call();
		((FileRepository) git.getRepository()).convertRefStorage("reftable", true, false);
		

		git.tag().setName("foo").call();
		System.out.println(git.getRepository().findRef("foo"));

		System.out.println("Now change the Tag via git bash (git tag --force foo v7.0.0.202408201547-m3)");
		System.out.println("and press enter. You will see, the tag in jgit doesn't change");
		Scanner scanner = new Scanner(System.in);
		scanner.next();
		
		System.out.println(git.getRepository().exactRef("refs/tags/foo"));
		
		git.close();
`

outputs:

`Ref[refs/tags/foo=ef0923a81c00465034705a24a30b4c8d6878729f(3)]
Now change the Tag via git bash (git tag --force foo v7.0.0.202408201547-m3)
and press enter. You will see, the tag in jgit doesn't change

Ref[refs/tags/foo=ef0923a81c00465034705a24a30b4c8d6878729f(3)]
`

For jgit, the ref didn't change

output from gitbash 

`$ git tag --force foo v7.0.0.202408201547-m3
Unlink of file 'C:/temp/jgit/.git/reftable/000000000003-000000000003-1b2b152a.ref' failed. Should I try again? (y/n) y
Unlink of file 'C:/temp/jgit/.git/reftable/000000000003-000000000003-1b2b152a.ref' failed. Should I try again? (y/n) y
Unlink of file 'C:/temp/jgit/.git/reftable/000000000003-000000000003-1b2b152a.ref' failed. Should I try again? (y/n) n
Unlink of file 'C:/temp/jgit/.git/reftable/000000000003-000000000003-1b2b152a.ref' failed. Should I try again? (y/n) n
Updated tag 'foo' (was ef0923a81)

$ git show-ref | grep foo
727666694e80ed39488451503d72f65bb4ef69f8 refs/tags/foo
`

you see #101 but the ref is changed

### Expected behavior

Changes on tables.list should call

ReftableDatabase.clearCache() **and**
FileReftableStack.reload()

in FileReftableDatabase

### Other information

I hope, this helps to improve jgit :)

**Comment by zotttelbart** [comment]
Hello,

the bug is still present in jgit 7.1 and git 2.47.1.

**Comment by msohn** [comment]
Pushed https://eclipse.gerrithub.io/c/eclipse-jgit/jgit/+/1206683 for review

**Comment by zotttelbart** [comment]
@msohn thx you. Its work!
Do you have any idea to solve #101?

**Comment by lucamilanesio** [comment]
> [@msohn](https://github.com/msohn) thx you. Its work!

See the [discussion on the original JGit RefTable Change 146568])(https://review.gerrithub.io/c/eclipse-jgit/jgit/+/146568) with @hanwen which brought some lights on the design behind the behaviour you are observing.

In a nutshell, when you create the repository object and the associated refDatabase (backed by a refTable), it takes a _consistent snapshot_ of the refs by reading the stack of immutable files and keeping them open. Whatever happens on the filesystem **is not** going to be detected _on the fly_ until you try to perform an update.

If the update happens _behind JGit's back_ but not inside the JGit's RefTable code, it won't be noticed until you close the repository and open it again.

Bottom line: I believe it is the way it works. Also it is consistent with the [RefTable specs of what readers should do](https://git-scm.com/docs/reftable#_readers).

> Do you have any idea to solve [#101](https://github.com/eclipse-jgit/jgit/issues/101)?

I believe that's unrelated to this problem and is more about compaction.

**Comment by zotttelbart** [comment]
> In a nutshell, when you create the repository object and the associated refDatabase (backed by a refTable), it takes a _consistent snapshot_ of the refs by reading the stack of immutable files and keeping them open. Whatever happens on the filesystem **is not** going to be detected _on the fly_ until you try to perform an update.
> 
> If the update happens _behind JGit's back_ but not inside the JGit's RefTable code, it won't be noticed until you close the repository and open it again.
> 

In Eclipse / egit its impossible to use repos with reftables. It leeds to corrupted filetables. With refdirectories (and packed-refs) its not a problem because, jgit refresh the cached refs.

**Comment by lucamilanesio** [comment]
> In Eclipse / egit its impossible to use repos with reftables. It leeds to corrupted filetables. 

I've actually performed some JGit + CGit tests with a repository using ref-table and ended up with a corrupted filetables as well. Using JGit only doesn't show any issues though. I suspect that's a different issue related to the C implementation of ref-tables.

> With refdirectories (and packed-refs) its not a problem because, jgit refresh the cached refs.

The two problems are separate: one thing is to _not notice_ the updates on the filesystem and another is to generate corrupted data.

With Gerrit, for example, the issue isn't really in JGit but [a bug in Gerrit](https://issues.gerritcodereview.com/issues/392541994) where the Repository reference counting was diverging hence keeping the ref-table open indefinitely. Once the underlying Gerrit issue is fixed, I cannot reproduce the stale ref-table anymore.

Maybe the same is possible with EGit?

**Comment by lucamilanesio** [comment]
@msohn I applied the [Gerrit fix on Repository reference-counting](https://gerrit-review.googlesource.com/c/gerrit/+/449083) and configured the JGit `RepositoryCache` to expire the repositories after 1 msec with the following:

```
[core]
	repositoryCacheExpireAfter = 1
	repositoryCacheCleanupDelay = 1
```

As a result, the JGit ref-table works as expected when used by Gerrit:
- ref-table files are released when repo isn't used
- older ref-table files are released when performing auto-compaction
- JGit can see the changes on the repository when performed outside Gerrit

Bottom line: there could be some similar problem in EGit with the reference counting that should be investigated IMHO.
The code reported in this bug report is actually working as documented in the [ref-table specifications for the reader](https://git-scm.com/docs/reftable#_readers).

@msohn  @hanwen  @zotttelbart, what do you think?

From Gerrit's perspective, all works without needing changes on JGit's implementation of the ref-table.

**Comment by zotttelbart** [comment]
I tested it in egit with version 7.1.0.

I do

git checkout in cgit 

eclipse didn't recognize the new branch. In commit view in egit all files which changed through the checkout are displayed as staged changes and it shows the old branch:

![Image](https://github.com/user-attachments/assets/63ab4093-3336-49ec-afff-7f564ef84d68)

With refdirectories all works expected and egit recognize the changed HEAD ref.

**Comment by msohn** [comment]
> [@msohn](https://github.com/msohn) I applied the [Gerrit fix on Repository reference-counting](https://gerrit-review.googlesource.com/c/gerrit/+/449083) and configured the JGit `RepositoryCache` to expire the repositories after 1 msec with the following:
> 
> ```
> [core]
> 	repositoryCacheExpireAfter = 1
> 	repositoryCacheCleanupDelay = 1
> ```
> 
> As a result, the JGit ref-table works as expected when used by Gerrit:
> 
> * ref-table files are released when repo isn't used
> * older ref-table files are released when performing auto-compaction
> * JGit can see the changes on the repository when performed outside Gerrit
> 
> Bottom line: there could be some similar problem in EGit with the reference counting that should be investigated IMHO. The code reported in this bug report is actually working as documented in the [ref-table specifications for the reader](https://git-scm.com/docs/reftable#_readers).
> 
> [@msohn](https://github.com/msohn) [@hanwen](https://github.com/hanwen) [@zotttelbart](https://github.com/zotttelbart), what do you think?

The time based eviction in RepositoryCache was introduced to fix a performance problem caused by immediately removing the repository's cached pack file pages from WindowCache when the repo's refcount drops to 0. This caused trashing on packfile contents cached in WindowCache which was immediately flushed from the cache without this time based eviction strategy.

> 
> From Gerrit's perspective, all works without needing changes on JGit's implementation of the ref-table.

**Comment by msohn** [comment]
https://eclipse.gerrithub.io/c/eclipse-jgit/jgit/+/1206683 was submitted.
For now you can configure `reftable.autorefresh = true` when using reftables in EGit.

**Comment by zotttelbart** [comment]
Thank you. I will test this with jgit 7.2

**Comment by zotttelbart** [comment]
@msohn I need Jgit 7.2 in maven central repository. When will it be available?
thx

**Comment by msohn** [comment]
thanks for the heads up, I just deployed it to maven central, replication until it's showing up there will take a bit

**Comment by msohn** [comment]
It's now published, see https://repo1.maven.org/maven2/org/eclipse/jgit/

**Comment by zotttelbart** [comment]
thx you. I think I will test this feature next week.

**Comment by syntonyze** [comment]
> thx you. I think I will test this feature next week.

@zotttelbart did you manage to test this by any chance? Can you share your findings?
